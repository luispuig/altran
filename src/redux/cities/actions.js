// @flow

import openWeather from "../../services/openWeather";

import type { _RequestPayload, _SuccesPayload, _FailPayload } from "./types.js"
import type { _State } from "../types.js"


export const CITY_UPDATE__REQUEST: string = "CITY_UPDATE__REQUEST";
export const CITY_UPDATE__FAIL: string = "CITY_UPDATE__FAIL";
export const CITY_UPDATE__SUCCESS: string = "CITY_UPDATE__SUCCESS";

export const cityUpdateAll = () => {

  /*
    ApiWeather allows to get all cities in one request with a limit of 20.
    But each city will be fetched individually for showing a more standard way.
    This way will allow future features like different update time per city.
   */

  return (dispatch:any => void, getState:any => void ) => {

    const state:_State = getState();
    const { cities } = state;
    // Dispatch Request Action for each City
    cities.map( item => dispatch( cityUpdate( { id: item.id } ) ) );
  }

};

export const cityUpdate = ( { id }:_RequestPayload, date:?Date = undefined ) => {
  return (dispatch:any => void) => {

    // Set City State to Fetching Request
    dispatch( cityUpdate_Request( { id } ) );

    // Fetch Promise
    return openWeather.getTempByCityId( id ).then(
        (temperature:number) => {
          dispatch( cityUpdate_Success  ( { id, temperature } ) );
        },
        (error:{ message:string }) => {
          dispatch( cityUpdate_Fail( { id, error: error.message } ) ) }
    );
  }
};

export const cityUpdate_Request = ({ id }:_RequestPayload) => ({
  type: CITY_UPDATE__REQUEST,
  payload: { id },
});

export const cityUpdate_Fail = ({id, error}:_FailPayload) => ({
  type: CITY_UPDATE__FAIL,
  payload: {id, error}
});

export const cityUpdate_Success = ({id, temperature}:_SuccesPayload) => ({
  type: CITY_UPDATE__SUCCESS,
  payload: {id, temperature}
});
