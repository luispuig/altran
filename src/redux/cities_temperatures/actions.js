// @flow
import type {_AddPayload} from "./types"

export const CITY_TEMPERATURE__ADD:string = "CITY_TEMPERATURE__ADD";

export const cityTemperature_Add = ({id, temperature, date}:{..._AddPayload, date:?Date}) => ({
  type: CITY_TEMPERATURE__ADD,
  payload: { id, temperature, date: (date !== undefined)? date: new Date() }
});
