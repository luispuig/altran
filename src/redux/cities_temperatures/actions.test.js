import * as actions from "./actions";

it('Creates Action cityTemperature_Add', () => {
  const action = actions.cityTemperature_Add({ id:3871336, date: new Date('2017-12-06T14:14:58.080Z'), temperature: 25 })
  expect(action).toMatchSnapshot();
});
