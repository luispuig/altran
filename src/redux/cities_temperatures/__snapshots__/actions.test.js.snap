// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creates Action cityTemperature_Add 1`] = `
Object {
  "payload": Object {
    "date": 2017-12-06T14:14:58.080Z,
    "id": 3871336,
    "temperature": 25,
  },
  "type": "CITY_TEMPERATURE__ADD",
}
`;
